---
title: "Analysis of GBM / random forest models"
output: html_document
---

This is a model test-bed, providing a large number of plotting/EDA scripts to analyse model performance. 

#Set-up
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

#Set-up and data import
library(dplyr)
library(tibble)
library(tidyr)
library(caret)
library(RColorBrewer)
library(ggrepel)
library(gtools)
library(Rmisc)
library(gbm)
library(stringr)
````



##Import
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
#setwd("~/Documents/nstringjul18")
load(file="allns_data.rdata")
load("vardat.rdata")
load("fullns_data.rdata")
load(file="qualitydatwnew.rdata")
load(file="sampleinfob.rdata")
load(file="markerchoiceinfo.rdata")
      
load(file="lena.rdata")
       
       
````

##Import model

This is simply for an example model (below). More models are available. 

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

load(file="gbmfullrsq3.rdata")

mymodel <- gbmfullrsqinc3

```

#Plots

##Plot model params

```{r gbm plot, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfob <- read.csv("sialplus.csv")



colnames(ctfull.norm) <- sampleinfob$sample[1:length(colnames(ctfull.norm))]

colnames(ctfull.unnorm) <- sampleinfob$sample[1:length(colnames(ctfull.unnorm))]

colnames(ctfull.intonly) <-  sampleinfob$sample[1:length(colnames(ctfull.norm))]

  
colnames(ctfull.hk) <-  sampleinfob$sample[1:length(colnames(ctfull.norm))]


```



```{r gbm plot, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
plot(mymodel)

#save(gbmagemodel, file="gbmnew.rdata")
#summary(currbest$finalModel)
````

##Plot age regression

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


sampleinfoc <- subset(sampleinfob, sampleinfob$sample %in% colnames(ctfull.norm))


markexp <-  data.frame( marker = colnames(t(ctfull.norm)), 
                        meanexp = colMeans(t(ctfull.norm), na.rm = T))








md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY") &
                 sampleinfoc$type %in% c("O", "Y", "M")
))

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]


#thresh <- 20

#md <- md[,colnames(md) %in% markexp$marker[markexp$meanexp > thresh]]


md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))


#md <- subset(md, md$sample != "Coh9")
md$sub <- ifelse(md$intrain, "train", "test")
md$sub[md$set == "new"] <- "validation"




RMSEtr <- sqrt(mean( (md$gbmagepred[md$sub == "train"] - md$predage[md$sub =="train"])   ^2))
RMSEte <- sqrt(mean( (md$gbmagepred[md$sub == "test"] - md$predage[md$sub =="test"])   ^2))
RMSEval <- sqrt(mean( (md$gbmagepred[md$sub == "validation"] - md$predage[md$sub =="validation"])   ^2))


(p=    ggplot(md, aes(x=predage, y=gbmagepred,colour=sub))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  geom_smooth(method="lm", colour="black")+
#  geom_point(data=subset(csn, csn$exp== "Size-age"), aes(x=predage, y=pred), colour="forest green")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3", "dark blue"), name="Data partition")+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3), "\nRMSE in validation data: ", round(RMSEval,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          
                        
                          
                          
                          ))
                          
                          
                          #as.character(mymodel$bestTune[1:5]))))
          
        
  
  )#+  facet_wrap(~intrain)







gbmrsq<-p


#save(gbmcompro, file="gbmcompro.rdata")

#ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("gbmold.pdf"), useDingbats=FALSE, limitsize = FALSE)
```



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfob <- read.csv("sialplus.csv")



sampleinfoc <- subset(sampleinfob, sampleinfob$sample %in% colnames(ctfull.norm))



#mymodel <- gbmfullrsqcompro4
#gbmfullrsqinc5

md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                  sampleinfoc$qual != "bad" &
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY")&#, "regen", "regen-long") &
                 sampleinfoc$type %in% c("O", "Y", "M")
))

#md2 <- as.data.frame(subset(t(ctfull.norm),                             sampleinfoc$sample %in% baduns))
                      

#md <- rbind(md, md2)
md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]




md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))


#md <- subset(md, md$sample != "Coh9")
md$sub <- ifelse(md$intrain, "train", "test")
md$sub[md$set == "new"] <- "validation"





RMSEtr <- sqrt(mean( (md$gbmagepred[md$intrain == T] - md$predage[md$intrain == T])   ^2))
RMSEte <- sqrt(mean( (md$gbmagepred[md$intrain == F] - md$predage[md$intrain == F])   ^2))


#(trgbmlm <- summary(lm(data=md[md$intrain,], predage~gbmagepred)))
#(tegbmlm <- summary(lm(data=md[md$intrain==F], predage~gbmagepred)))
#round(trgbmlm$r.squared, 3)

(p=    ggplot(md, 
  aes(x=predage, y=gbmagepred,colour=intrain))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  geom_smooth(data=subset(md, md$intrain), method="lm", se=F)+
   theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age \n Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used ",
                          "\nTraining data:   RMSE: ", round(RMSEtr,3),#, "         R-squared: ",round(trgbmlm$r.squared, 3),
                          "\nTest data:         RMSE: ", round(RMSEte,3)#, "     R-squared: ",round(tegbmlm$r.squared, 3)
                          )
          )+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          ))
  
  )



ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("gbmfullqualinc3.pdf"), useDingbats=FALSE, limitsize = FALSE)
```


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}




```


#Regen effect as boxplot

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY", "regen", "regen-long") 
))

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]

md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))

md$group <- case_when(md$intrain == T ~ "Training Set", 
                      md$intrain == F & md$exp %in% c("Size-age", "cohorts", "OvY") ~ "Test Set", 
                      md$intrain == F & md$exp %in% c("regen", "regen-long") & md$reg == "" ~ "Unamputated",
                      md$intrain == F & md$exp %in% c("regen", "regen-long") & md$reg %in% c("R", "R-A") ~ "Regenerated"
)



#md$group <- case_when(md$intrain == T ~ "Training Set", 
#                      md$intrain == F & md$exp %in% c("regen", "regen-long") & md$reg %in% c("R", "R-A") ~ "Regenerated",
#                      md$intrain == F & md$exp %in% c("Size-age", "cohorts", "OvY", "regen", "regen-long") ~ "Test Set"
#)


mhk2 <- mhk
mhk2$total <- rowSums(mhk2)


mhk2 <- rownames_to_column(mhk2 , var= "sample")




md <- left_join(md, mhk2)




md$resid <- md$predage - md$gbmagepred

(p= ggplot(md, aes(x= group, y=resid, fill=group)) +
  geom_boxplot()+
  theme_bw()+ 
  xlab("" ) +
# scale_fill_manual(values=c("forest green", "cornflower blue", "red3"), name="Sample type")+
  ylab("Residual of GBM age vs length-based age (weeks) \n(Negative values indicated regeneration)")
  
)



ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("regeffbox.pdf"), useDingbats=FALSE, limitsize = FALSE)
```

#Regression variant plot

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

(p1=
   ggplot(subset(md, md$group == "Training Set"), aes(x=predage, y=gbmagepred,colour=group, alpha= group))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_colour_manual(values=c("red3"), name="Sample type", guide=F)+
  scale_alpha_manual(values= c(1), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  geom_smooth(data=subset(md, md$group == "Training Set"), method="lm", se=F)+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


#ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("trainonly.png"),  bg = "transparent",limitsize = FALSE)

```

#Regression variant

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md$group[md$exp %in% c("regen", "regen-long") & md$reg == ""]  <- "Contralateral"

md$exp %in% c("Size-age", "cohorts", "OvY")


md1 <- subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") | md$sample %in% baduns)


(p1=
   ggplot(md1, aes(x=predage, y=gbmagepred,colour=group, alpha= group))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_colour_manual(values=c("grey", "cornflower blue", "red3"), name="Sample type", guide=F)+
  scale_alpha_manual(values= c(0, 1,1), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  geom_smooth(data=subset(md1, md1$intrain), method="lm", se=F)+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


#ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("traintest.png"),  bg = "transparent",limitsize = FALSE)

```

##Regen effect

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md$group[md$exp %in% c("regen", "regen-long") & md$reg == ""]  <- "Contralateral"

#md$group[md$type %in% c("MR-A", "OR-A")] <- "~12 weeks regenerated"

(p1=
   ggplot(md, aes(x=predage, y=gbmagepred,colour=group, alpha= group))+
  geom_smooth(data=subset(md, md$intrain & md$exp!= "regen"), method="lm", se=F)+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_colour_manual(values=c("grey", "forest green", "cornflower blue", "red3"), name="Sample type", guide=F)+
  scale_alpha_manual(values= c(1, 1,0,0), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("reg.png"),  bg = "transparent",limitsize = FALSE)



```

#Regeneration effect

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md$group[md$exp %in% c("regen", "regen-long") & md$reg == ""]  <- "Contralateral"

#md$group[md$type %in% c("MR-A", "OR-A")] <- "~12 weeks regenerated"

(p1=
   ggplot(md, aes(x=predage, y=gbmagepred,fill=group,colour=intrain, alpha= group))+
  geom_smooth(data=subset(md, md$intrain & md$exp!= "regen"),colour="red3", method="lm", se=F)+
  geom_point(shape=21)+
  scale_colour_manual(values= c("black", "white"),guide=F)+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_fill_manual(values=c("grey", "forest green", "cornflower blue", "red3"), name="Sample type", guide=F)+
  scale_alpha_manual(values= c(1, 1,0,0), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("reg.png"),  bg = "transparent",limitsize = FALSE)

```




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfoc <- subset(sampleinfob, sampleinfob$sample %in% colnames(ctfull.norm))






md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY", "regen") 
))

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]

md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))


#md <- subset(md, md$sample != "Coh9")
md$sub <- ifelse(md$intrain, "train", "test")
md$sub[md$set == "new"] <- "validation"

mdr <- subset(md, md$exp == "regen") #%>%
#  select(sample, predage, gbmagepred)


reg <- subset(mdr, str_sub(mdr$sample,-3,-1) == "reg")  %>%
  select(sample, gbmagepred, predage )
names(reg)[names(reg)=="gbmagepred"] <- "gbmaagepredreg"
reg$sample <- substr(reg$sample, 1, str_length(reg$sample)-3)

un <- subset(mdr, str_sub(mdr$sample,-2,-1) == "un")  #%>%
#  select(sample, gbmagepred )
un$sample <- substr(un$sample, 1, str_length(un$sample)-2)

unvreg <- left_join(un, reg)
unvreg$dir <- unvreg$gbmaagepredreg < unvreg$gbmagepred
unvreg$diff <-   unvreg$gbmagepred - unvreg$gbmaagepredreg

(
reg<-ggplot(unvreg, aes(x= predage, 
                   xend=predage,  
                   y= gbmagepred, 
                   yend= gbmaagepredreg, 
                   colour= dir
                   )) +
  geom_segment() +
  facet_wrap(~set)+
  theme_bw()

)




```






##Variable importance

Variable importance obtained via permutation 

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


#mymodel <- currbestgbm
#mymodel <- gbmcompro
vrimpgbm <- summary(mymodel)


vrimpgbm$sname <- substr(vrimpgbm$var,1,2)
vrimpgbm$sname <- factor(vrimpgbm$sname, levels=vrimpgbm$sname[order(vrimpgbm$rel.inf, decreasing=T)])

(p=ggplot(vrimpgbm, aes(x=sname, y=rel.inf))+
  theme_bw()+
  geom_bar(stat="identity", fill="red3")+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("Marker")+
  ylab("Contribution to model")+
  ggtitle("Marker contributions", subtitle= paste0("GBM model built with ", length(mymodel$coefnames), " markers"))
)
top50 <- as.character(vrimpgbm$var[order(vrimpgbm$rel.inf, decreasing =T)][1:50])

contriblist<- as.character(vrimpgbm$var)[vrimpgbm$rel.inf >0]

gvar<- as.character(vrimpgbm$var)[vrimpgbm$rel.inf >0.0023]


````

##Gene expression profiles for model features

This is a huge plot, and is exported externally

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.height = 12}



md <- as.data.frame(subset(t(ctfull.norm), sampleinfob$sex == "F" &
#               sampleinfo$qual == "ok"&
                 sampleinfob$prep== "Luke"&
                 sampleinfob$exp %in% c("Size-age", "cohorts", "OvY") &
 #                sampleinfo$ctg>2 &
                sampleinfob$codeset == "phaw_1" &
                 sampleinfob$type %in% c("O", "Y", "M")
))

md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]

#md <- subset(md, rownames(md) %in% sampleinfob$sample[!is.na(sampleinfob$predage)])

md <-  as.data.frame(md/rowMeans(md))

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
#md <- md %>%
#  rownames_to_column(var="sample") %>%
#  left_join(sampleinfob,  by="sample")
md <- md %>%
#  rownames_to_column(var="sample") %>%
  left_join(sampleinfob,  by="sample") %>%
  gather(key=marker, value= exp, mymodel$coefnames)%>%
  left_join(minf, by=c("marker"="name"))%>%
  left_join(select(vrimpgbm, -sname), by= c("marker"="var"))

md$lab <- paste0(md$sname, ": ", round(md$rel.inf,3))


md$lab <- factor(md$lab, levels=unique(md$lab[order(md$rel.inf, decreasing=T)]))

(p=ggplot(md, aes(x=predage, y= exp, colour=dir))+
  geom_smooth()+
  geom_point(size=0.5)+
  theme_bw()+
  facet_wrap(~lab, scales="free", ncol=3)+
  scale_colour_brewer(palette="Set1")+
  xlab("Age (based on body length)")+
  ylab("Normalised expression")+
  theme(legend.direction = 'horizontal', legend.position = 'bottom')+
 
  ggtitle("Marker profiles", subtitle= paste0("markers from GBM built with ", length(mymodel$coefnames), " markers")) )





ggsave(plot=p,height=18,width=6,dpi=200, filename=paste("gbmcontribprofiles.pdf"), useDingbats=FALSE, limitsize = FALSE)


```


