---
title: "Analysis of GBM / random forest models"
output: html_document
---




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

#Set-up and data import
library(dplyr)
library(tibble)
library(tidyr)
library(caret)
library(RColorBrewer)
library(ggrepel)
library(gtools)
library(Rmisc)
library(gbm)
library(stringr)
````




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
#setwd("~/Documents/nstringjul18")
load(file="allns_data.rdata")
load("vardat.rdata")
load("fullns_data.rdata")
load(file="qualitydat.rdata")
load(file="qualitydatwnew.rdata")
load(file="sampleinfob.rdata")
load(file="markerchoiceinfo.rdata")
load(file="gbmrsqopt.rdata")
      
load(file="lena.rdata")
load(file="gbmrmseopt.rdata") 
       
       
````



#Plot model params

```{r gbm plot, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfob <- read.csv("sialplus.csv")



colnames(ctfull.norm) <- sampleinfob$sample[1:length(colnames(ctfull.norm))]

colnames(ctfull.unnorm) <- sampleinfob$sample[1:length(colnames(ctfull.unnorm))]

colnames(ctfull.intonly) <-  sampleinfob$sample[1:length(colnames(ctfull.norm))]

  
colnames(ctfull.hk) <-  sampleinfob$sample[1:length(colnames(ctfull.norm))]


```



```{r gbm plot, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
plot(mymodel)

#save(gbmagemodel, file="gbmnew.rdata")
#summary(currbest$finalModel)
````

#Plot age regression
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfob <- read.csv("sialplus.csv")



sampleinfoc <- subset(sampleinfob, sampleinfob$sample %in% colnames(ctfull.norm))


markexp <-  data.frame( marker = colnames(t(ctfull.norm)), 
                        meanexp = colMeans(t(ctfull.norm), na.rm = T))




mymodel <- gbmfullrsqinc





md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY") &
 #                sampleinfo$ctg>2 &
                 sampleinfoc$type %in% c("O", "Y", "M")
))

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))


#md <- subset(md, md$sample != "Coh9")
md$sub <- ifelse(md$intrain, "train", "test")
md$sub[md$set == "new"] <- "validation"




RMSEtr <- sqrt(mean( (md$gbmagepred[md$sub == "train"] - md$predage[md$sub =="train"])   ^2))
RMSEte <- sqrt(mean( (md$gbmagepred[md$sub == "test"] - md$predage[md$sub =="test"])   ^2))
RMSEval <- sqrt(mean( (md$gbmagepred[md$sub == "validation"] - md$predage[md$sub =="validation"])   ^2))


(p=    ggplot(md, aes(x=predage, y=gbmagepred,colour=sub))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  geom_smooth(method="lm", colour="black")+
#  geom_point(data=subset(csn, csn$exp== "Size-age"), aes(x=predage, y=pred), colour="forest green")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3", "dark blue"), name="Data partition")+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3), "\nRMSE in validation data: ", round(RMSEval,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          
                        
                          
                          
                          ))
                          
                          
                          #as.character(mymodel$bestTune[1:5]))))
          
        
  
  )#+  facet_wrap(~intrain)







gbmrsq<-p


#save(gbmcompro, file="gbmcompro.rdata")

#ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("gbmold.pdf"), useDingbats=FALSE, limitsize = FALSE)
```

```{r Gradient boosting build, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
uns<-  sampleinfoc$sample[
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                  sampleinfoc$reg == "" &
                 sampleinfoc$exp %in% c("regen", "regen-long") &
                 sampleinfoc$type %in% c("O", "Y", "M")
] %>%
  as.character()

uns <- uns[grepl("un", uns)]


names <- substr(uns,1,nchar(uns)-2)


regs<-  sampleinfoc$sample[
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                  sampleinfoc$reg %in% c("R", "R-A") &
                 sampleinfoc$exp %in% c("regen", "regen-long") 
] %>%
  as.character()

badregs <- qualsum$sample[
  qualsum$sample %in% regs &
  qualsum$good == F
] %>%
  as.character()

baduns <- paste0(substr(badregs,1,nchar(badregs)-3), "un")


#regs <- substr(regs,1,nchar(regs)-2)





````


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfob <- read.csv("sialplus.csv")



sampleinfoc <- subset(sampleinfob, sampleinfob$sample %in% colnames(ctfull.norm))



mymodel <- gbmfullrsqinc13
#gbmfullrsqinc5

md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                  sampleinfo$qual != "bad" &
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY")&#, "regen", "regen-long") &
                 sampleinfoc$type %in% c("O", "Y", "M")
))

md2 <- as.data.frame(subset(t(ctfull.norm),                             sampleinfoc$sample %in% baduns))
                      

md <- rbind(md, md2)

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]



md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))


#md <- subset(md, md$sample != "Coh9")
md$sub <- ifelse(md$intrain, "train", "test")
md$sub[md$set == "new"] <- "validation"





RMSEtr <- sqrt(mean( (md$gbmagepred[md$intrain == T] - md$predage[md$intrain == T])   ^2))
RMSEte <- sqrt(mean( (md$gbmagepred[md$intrain == F] - md$predage[md$intrain == F])   ^2))


#(trgbmlm <- summary(lm(data=md[md$intrain,], predage~gbmagepred)))
#(tegbmlm <- summary(lm(data=md[md$intrain==F], predage~gbmagepred)))
#round(trgbmlm$r.squared, 3)

(p=    ggplot(md, 
  aes(x=predage, y=gbmagepred,colour=intrain))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  geom_smooth(data=subset(md, md$intrain), method="lm", se=F)+
   theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age \n Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used ",
                          "\nTraining data:   RMSE: ", round(RMSEtr,3),#, "         R-squared: ",round(trgbmlm$r.squared, 3),
                          "\nTest data:         RMSE: ", round(RMSEte,3)#, "     R-squared: ",round(tegbmlm$r.squared, 3)
                          )
          )+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          ))
  
  )



ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("gbmfullqualinc3.pdf"), useDingbats=FALSE, limitsize = FALSE)
```


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}




```




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY", "regen", "regen-long") 
))

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]

md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))

md$group <- case_when(md$intrain == T ~ "Training Set", 
                      md$intrain == F & md$exp %in% c("Size-age", "cohorts", "OvY") ~ "Test Set", 
                      md$intrain == F & md$exp %in% c("regen", "regen-long") & md$reg == "" ~ "Unamputated",
                      md$intrain == F & md$exp %in% c("regen", "regen-long") & md$reg %in% c("R", "R-A") ~ "Regenerated"
)



md$group <- case_when(md$intrain == T ~ "Training Set", 
                      md$intrain == F & md$exp %in% c("regen", "regen-long") & md$reg %in% c("R", "R-A") ~ "Regenerated",
                      md$intrain == F & md$exp %in% c("Size-age", "cohorts", "OvY", "regen", "regen-long") ~ "Test Set"
)


mhk2 <- mhk
mhk2$total <- rowSums(mhk2)


mhk2 <- rownames_to_column(mhk2 , var= "sample")




md <- left_join(md, mhk2)




md$resid <- md$predage - md$gbmagepred

(p= ggplot(md, aes(x= group, y=resid, fill=group)) +
  geom_boxplot()+
  theme_bw()+ 
  xlab("" ) +
# scale_fill_manual(values=c("forest green", "cornflower blue", "red3"), name="Sample type")+
  ylab("Residual of GBM age vs length-based age (weeks) \n(Negative values indicated regeneration)")
)



ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("regeffbox.pdf"), useDingbats=FALSE, limitsize = FALSE)
```
#presimg 1
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

(p1=
   ggplot(md, aes(x=predage, y=gbmagepred,colour=group, alpha= group))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_colour_manual(values=c("forest green", "cornflower blue", "red3"), name="Sample type", guide=F)+
  scale_alpha_manual(values= c(0, 0,1), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  geom_smooth(data=subset(md, md$intrain), method="lm", se=F)+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("trainonly.png"),  bg = "transparent",limitsize = FALSE)

```
#presimg 2

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md1 <- subset(md,md$exp %in% c("Size-age", "cohorts", "OvY"))

(p1=
   ggplot(md1, aes(x=predage, y=gbmagepred,colour=group, alpha= group))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_colour_manual(values=c("forest green", "cornflower blue", "red3"), name="Sample type", guide=F)+
#  scale_alpha_manual(values= c(0, 1,1), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  geom_smooth(data=subset(md, md$intrain), method="lm", se=F)+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("traintest.png"),  bg = "transparent",limitsize = FALSE)

```
#presimg3

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md$group[md$exp %in% c("regen", "regen-long") & md$reg == ""]  <- "Contralateral"

md$group[md$type %in% c("MR-A", "OR-A")] <- "~12 weeks regenerated"

(p1=
   ggplot(md, aes(x=predage, y=gbmagepred,colour=group, alpha= group))+
  geom_smooth(data=subset(md, md$intrain), method="lm", se=F)+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  theme_bw()+
  scale_colour_manual(values=c("olivedrab3", "grey", "forest green", "cornflower blue", "red3"), name="Sample type", guide=F)+
  scale_alpha_manual(values= c(1,1, 1,0,0), guide=F)+
   ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Criteria: 4 housekeeping genes over 3 reads, Chi2 < 500", "\nModel type: ", 
                          toupper(mymodel$method), ", ", 
                          length(mymodel$coefnames), 
                          " markers used \nRMSE in training data: ", 
                          round(RMSEtr,3), 
                          "\nRMSE in test data: ", 
                          round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
 ))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

)


ggsave(plot=p1,height=8,width=12,dpi=300, device= "png", filename=paste("reg.png"),  bg = "transparent",limitsize = FALSE)

```




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

sampleinfoc <- subset(sampleinfob, sampleinfob$sample %in% colnames(ctfull.norm))




mymodel <- gbmfullrmse3

md <- as.data.frame(subset(t(ctfull.norm), 
                sampleinfoc$sex == "F" &              
                  sampleinfoc$prep== "Luke"&
                 sampleinfoc$exp %in% c("Size-age", "cohorts", "OvY", "regen") 
))

md <-subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]

md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)

md <- as.data.frame(md) %>%
  rownames_to_column(var ="sample") %>%
  left_join(sampleinfoc, by="sample")

predage <- predict(polymod3, newdata=md)

md$predage <- predage
md <- subset(md, !(is.na(md$predage)))


#md <- subset(md, md$sample != "Coh9")
md$sub <- ifelse(md$intrain, "train", "test")
md$sub[md$set == "new"] <- "validation"

mdr <- subset(md, md$exp == "regen") #%>%
#  select(sample, predage, gbmagepred)


reg <- subset(mdr, str_sub(mdr$sample,-3,-1) == "reg")  %>%
  select(sample, gbmagepred, predage )
names(reg)[names(reg)=="gbmagepred"] <- "gbmaagepredreg"
reg$sample <- substr(reg$sample, 1, str_length(reg$sample)-3)

un <- subset(mdr, str_sub(mdr$sample,-2,-1) == "un")  #%>%
#  select(sample, gbmagepred )
un$sample <- substr(un$sample, 1, str_length(un$sample)-2)

unvreg <- left_join(un, reg)
unvreg$dir <- unvreg$gbmaagepredreg < unvreg$gbmagepred
unvreg$diff <-   unvreg$gbmagepred - unvreg$gbmaagepredreg

(
reg<-ggplot(unvreg, aes(x= predage, 
                   xend=predage,  
                   y= gbmagepred, 
                   yend= gbmaagepredreg, 
                   colour= dir
                   )) +
  geom_segment() +
  facet_wrap(~set)+
  theme_bw()

)




```





```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



              
md <- subset(md, md$exp != "regen")

RMSEtr <- sqrt(mean( (md$gbmagepred[md$sub == "train"] - md$predage[md$sub =="train"])   ^2))
RMSEte <- sqrt(mean( (md$gbmagepred[md$sub == "test"] - md$predage[md$sub =="test"])   ^2))
RMSEval <- sqrt(mean( (md$gbmagepred[md$sub == "validation"] - md$predage[md$sub =="validation"])   ^2))





(p=    ggplot(md, aes(x=predage, y=gbmagepred,colour=sub))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression")+
  geom_smooth(method="lm", colour="black")+
#  geom_point(data=subset(csn, csn$exp== "Size-age"), aes(x=predage, y=pred), colour="forest green")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3", "dark blue"), name="Data partition")+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3), "\nRMSE in validation data: ", round(RMSEval,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          
                        
                          
                          
                          ))
                          
                          
                          #as.character(mymodel$bestTune[1:5]))))
          
        
  
  )#+  facet_wrap(~intrain)







p3<-p


#save(gbmcompro, file="gbmcompro.rdata")

#ggsave(plot=p,height=8,width=12,dpi=200, filename=paste("gbmold.pdf"), useDingbats=FALSE, limitsize = FALSE)
```


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
md <- subset(md, is.na(md$wkage) ==F)


RMSEtr <- sqrt(mean( (md$gbmagepred[md$intrain == T] - as.numeric(md$wkage[md$intrain == T]) )  ^2))
RMSEte <- sqrt(mean( (md$gbmagepred[md$intrain == F] -  as.numeric(md$wkage[md$intrain == F]))   ^2))


(p=    ggplot(md, aes(x=predage, y=wkage,colour=intrain))+
  geom_point()+
  xlab("Age (weeks) cohort age data") +
  ylab("Age (weeks) from marker gene expression")+
  geom_smooth(method="lm", colour="black")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          
                        
                          
                          
                          ))
                          
                          
                          #as.character(mymodel$bestTune[1:5]))))
          
        
  
  )

```


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



pdf("gbmcomp", height=12, width=16)
multiplot(p1, p2, p3, p4, cols=2)
dev.off()
```
#Plot residuals
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
md$resid <-md$gbmagepred- md$predage 

inres <- md$resid[md$intrain == T]


outres <- md$resid[md$intrain == F]


(p=    ggplot(md, aes(x=predage, y=resid,colour=intrain))+
  geom_bar(stat='identity')+
  xlab("Age (weeks) from body length") +
  ylab("Residual")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          ))
)


rest <- t.test(outres)
mtres <- mean(outres)


(p=    ggplot(md, aes(y=resid,fill=intrain))+
  geom_boxplot()+
  xlab("Age (weeks) from body length") +
  ylab("Residual")+
  theme_bw()+
 scale_fill_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
 
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used \n Mean residual (test):",round(mtres,2),"\n residual one-way t-test pval:", round(rest$p.value, 3)))+ 
    labs(caption = paste0("final parameters: ", 
                          names(mymodel$bestTune)[1], ": ", mymodel$bestTune[1], " ", 
                          names(mymodel$bestTune)[2], ": ", mymodel$bestTune[2], " ", 
                          names(mymodel$bestTune)[3], ": ", mymodel$bestTune[3], "  ", 
                          names(mymodel$bestTune)[4], ": ", mymodel$bestTune[4], ""
                          ))
)
```

####Variable importance

Variable importance obtained via permutation 

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


mymodel <- currbestgbm
mymodel <- gbmcompro
vrimpgbm <- summary(mymodel)


vrimpgbm$sname <- substr(vrimpgbm$var,1,2)
vrimpgbm$sname <- factor(vrimpgbm$sname, levels=vrimpgbm$sname[order(vrimpgbm$rel.inf, decreasing=T)])

(p=ggplot(vrimpgbm, aes(x=sname, y=rel.inf))+
  theme_bw()+
  geom_bar(stat="identity", fill="red3")+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("Marker")+
  ylab("Contribution to model")+
  ggtitle("Marker contributions", subtitle= paste0("GBM model built with ", length(mymodel$coefnames), " markers"))
)
top50 <- as.character(vrimpgbm$var[order(vrimpgbm$rel.inf, decreasing =T)][1:50])

contriblist<- as.character(vrimpgbm$var)[vrimpgbm$rel.inf >0]

gvar<- as.character(vrimpgbm$var)[vrimpgbm$rel.inf >0.0023]


````


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.height = 12}


md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
#               sampleinfo$qual == "ok"& 
                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") & 
 #                sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))



md <- md[,colnames(md) %in% mymodel$coefnames]
md <- subset(md, rownames(md) %in% sampleinfo$sample[!is.na(sampleinfo$predage)])

md <-  as.data.frame(md/rowMeans(md))

md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample") %>%
  gather(key=marker, value= exp, mymodel$coefnames)%>%
  left_join(minf, by=c("marker"="name"))%>%
  left_join(select(vrimpgbm, -sname), by= c("marker"="var"))

md$lab <- paste0(md$sname, ": ", round(md$rel.inf,3))


md$lab <- factor(md$lab, levels=unique(md$lab[order(md$rel.inf, decreasing=T)]))

(p=ggplot(md, aes(x=predage, y= exp, colour=dir))+
  geom_smooth()+
  geom_point(size=0.5)+
  theme_bw()+
  facet_wrap(~lab, scales="free", ncol=3)+
  scale_colour_brewer(palette="Set1")+
  xlab("Age (based on body length)")+
  ylab("Normalised expression")+
  theme(legend.direction = 'horizontal', legend.position = 'bottom')+
 
  ggtitle("Marker profiles", subtitle= paste0("markers from random forest model built with ", length(mymodel$coefnames), " markers")) )





ggsave(plot=p,height=18,width=6,dpi=200, filename=paste("gbmcontribprofiles.pdf"), useDingbats=FALSE, limitsize = FALSE)


```




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


vrimp <- left_join(vrimpgbm, vrimprf, by="sname")

ggplot(vrimp, aes(x=rel.inf,y=Overall, label=sname))+
  theme_bw()+
  ylab("Random Forest importance")+
  xlab("GBM relative influence")+
  scale_y_continuous(trans="log2")+
  scale_x_continuous(trans="log2")+
#  scale_y_log10()+
#  scale_x_log10()+
  geom_point(size=3, colour="red2")+
  geom_text_repel()+
  ggtitle("Marker importance: GBM vs Random Forest", subtitle= paste0("Same ", length(mymodel$coefnames), " markers"))


````





```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
#               sampleinfo$qual == "ok"&
                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") &
 #                sampleinfo$ctg>2 &
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")
))

md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]


md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample")


md$gbmagepred <- predict(mymodel, newdata=md)


md$intrain <- md$sample %in% rownames(mymodel$trainingData)

md$ageresid <- md$wkage - md$predage

md$gbmresid <- md$predage - md$gbmagepred

ggplot(md, aes(md$ageresid,  md$gbmresid ) )+
  geom_point(aes(colour= intrain))+
  geom_smooth(method="lm", se=F, colour="black")+
  theme_bw()++
#  scale_colour_brewer(palette="Set1")+
  ylab("Residual variation in GBM (exp v predage)")+
  xlab("Residual variation in polynomial (len v age)")+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  ggtitle("")


```


