---
title: "Lab meeting"
author: "Luke Hayden"
date: "14th November 2018"
output:
  beamer_presentation:
    theme: "Darmstadt"
    colortheme: "fly"
    fonttheme: "default"
slide_level: 3
---



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(reshape2)
library(ggbiplot)
library(tibble)
library(tidyr)
library(caret)
library(RColorBrewer)
library(ggrepel)
library(gtools)
library(FinCal)
library(ggrepel)

library(dplyr)
#library(e10)
````



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
#setwd("~/Documents/nstringjul18/nstringfull")
load(file="allns_data.rdata")
load(file="modelrf.rdata")
load(file="modelgbm.rdata")
load(file="hkdat.rdata")
load(file="sampleinfo.rdata")
load(file="qualitydat.rdata")
load(file="vardat.rdata")
load(file="goodgroups.rdata")
load(file="markerchoiceinfo.rdata")
load(file="gbmnew.rdata")
load(file="lena.rdata")
load(file="progmodels.rdata")
load(file="rfagemodelbadcv.rdata")


````

#Did everybody get a bun?



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
buns <- data.frame(Status=c("With bun", "Without bun"), Happiness=c(100, 3))

ggplot(buns, aes(Status, Happiness, fill=Status))+
  geom_bar(stat="identity")+
  theme_bw()+
  scale_fill_manual(values=c("cornflower blue", "red3"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
  



```


#Project

Examine the effect of regeneration on the molecular age profile of *Parhyale* limbs

##Designing codeset

*Nanostring as method to quantify gene expression

*200 genes in codeset

-195 genes chosen on the basis of differential expression analysis 

-5 control genes: do not vary in expression between conditions

#Old vs Young

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- t(subset(t(ctall.norm), sampleinfo$sex == "F" &
  #              sampleinfo$qual == "ok"& 
                 sampleinfo$exp %in% c("OvY", "Size-age", "cohorts") & 
                 sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))

si <- subset(sampleinfo, sampleinfo$sample %in% colnames(md))
old <- t(subset(t(md), si$age == "O"))
young <- t(subset(t(md), si$age == "Y"))
#regen <- t(subset(t(ctg1to9), sampleinfo$type == "OR" & sampleinfo$qual != "bad"))





nmd <- as.data.frame(md/rowMeans(md))


ns <- colnames(md)

nmd$name <- rownames(md)
nmd$sname <- substr(nmd$name, 1,2)

oy <- cbind(old, young)
t.result <-  apply(oy, 1, function (x) t.test(x[1:ncol(old)],x[ncol(old)+1: ncol(oy)]))



nmd$p_value <- unlist(lapply(t.result, function(x) x$p.value))

nmd <- left_join(nmd, geneinf, by="name")






nmd$YvOl2fc <- foldchange2logratio(foldchange(rowMeans(young), rowMeans(old)))


nmd$sname <- factor(nmd$sname, levels=nmd$sname[order(nmd$YvOl2fc)])
nmd$good <- nmd$p_value <0.05
nmd$di <- paste0(ifelse(nmd$dir == "upregulated with age", "up", "down"), ": ", nmd$sname)


sl <- as.character(sampleinfo$sample[sampleinfo$sample %in% colnames(nmd)])

mmd <- gather(as.data.frame(md), key="sample", value="expression", sl)

mnmd <- gather(nmd, key="sample", value="normexp", sl) %>%
  left_join(sampleinfo, by= "sample") %>%
  left_join(mmd, by="sample")


ggplot(mnmd, aes(x= sname, y=normexp, colour=type))+
#  geom_point()+
  theme_bw()+
  geom_boxplot(outlier.shape=NA)+
  scale_y_continuous(trans="log2")+
  facet_grid(~dir, space="free",scales="free" )+
#  geom_boxplot(outlier.shape = NA)+
  scale_colour_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  scale_fill_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  ylab("Normalised Expression") +
  xlab('Marker')+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
  


````


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

mlena <- gather(lena, key="model", value="pred", c("bcpred", "linpred", "polypred2", "polypred3"))

# 
# ggplot(mlena, aes(x=pred, y=length, colour=model))+
#   geom_smooth(aes(y=length, x=wkage), colour="orange")+
#   geom_line(size=1, alpha=0.6) +
#   theme_bw()+
#   geom_point(aes(x=wkage), colour="black")+
#   scale_colour_brewer(palette="Set1", labels= c("linear model" ,
#                  "Box-cox transformed linear model", 
#                  "2rd-degree polynomial",
#                  "3rd-degree polynomial")
#                         )+
#   xlab("Age (weeks)")+
#   ylab("Body length (mm)")+
#   ggtitle(label= "Linear regression age vs length", 
#           subtitle=paste0("linear model R-squared: ", round(summary(m)$r.squared, 3), 
#                  "\nBox-cox transformed linear model R-squared: ", round(summary(mnew)$r.squared, 3), 
#                  "\n2rd-degree polynomial R-squared: ", round(summary(polymod2)$r.squared, 2),
#                  "\n3rd-degree polynomial R-squared: ", round(summary(polymod3)$r.squared, 3)))
  
mlena$model <- factor(mlena$model, levels= c( "linpred", "bcpred","polypred2", "polypred3"))


variable_labeller <- function(variable,value){
  return(variable_names[value])
}


variable_names <- list( "linpred"=                   paste("linear model\n R-squared: ", round(summary(m)$r.squared, 3)), 
 "bcpred"=                 paste("Box-cox transformed linear model\n R-squared: ", round(summary(mnew)$r.squared, 3)), 
  "polypred2"=                paste("2rd-degree polynomial\n R-squared: ", round(summary(polymod2)$r.squared, 2)),
    "polypred3" =              paste("3rd-degree polynomial\n R-squared: ", round(summary(polymod3)$r.squared, 3))
                                     )



# 
# 
# ggplot(mlena, aes(x=pred, y=length, colour=model))+
#   geom_smooth(aes(y=length, x=wkage), colour="orange")+
#   geom_line(size=1, alpha=0.6) +
#   theme_bw()+
#   facet_wrap(~model, labeller=variable_labeller)+
#   geom_point(aes(x=wkage), colour="black")+
#   scale_colour_brewer(palette="Set1", guide=F)+
#   xlab("Age (weeks)")+
#   ylab("Body length (mm)")+
#   ggtitle("Age vs length models", subtitle= 
#             "Models compared with LOESS (orange)")
  
  
  
```


#Young vs old separation: PCA old vs young

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



log <- sampleinfo$reg == "" & 
  !(sampleinfo$age %in% c("N")) & 
  sampleinfo$sex == "F"&
  sampleinfo$exp %in% c("OvY", "Size-age") & 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8")) &
 sampleinfo$over10reads > 100 &
  sampleinfo$ctg > 2 &
  sampleinfo$prep == "Luke"

bcmin <- as.matrix(ctallgood.norm[,log== TRUE])
bcmin.groups <- subset(sampleinfo$type, log==TRUE)

tbc <- as.matrix(t(bcmin))

bc.pca <- prcomp(tbc,center = TRUE,scale. = TRUE) 
bc.groups <- bcmin.groups


ggbiplot(bc.pca, obs.scale = 1, var.scale = 1, ellipse = FALSE, circle = TRUE, var.axes=F, labels=rownames(tbc), groups=bc.groups)+ggtitle("PCA of Old vs Young (females): 50 chosen markers")+
    theme_bw() +
#  scale_colour_manual(values=c("cornflower blue", "red3", "black"))+
  theme(plot.title=element_text(size=8,face="bold"))+
#  geom_label(label=rownames(tbc), aes(colour=bc.groups))+
  scale_colour_manual(values= c("darkviolet","cornflower blue",  "red2"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )


````



#Expression/length relationship


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
  #              sampleinfo$qual == "ok"& 
#                 sampleinfo$exp %in% c("OvY", "Size-age", "cohorts") & 
                 sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")))
 # !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))



markerlist <- colnames(md)

md$sample <- rownames(md)


md <- left_join(md, select(sampleinfo, sample, predage), by="sample") %>%
  filter(!(is.na(predage))) %>%
  column_to_rownames(var="sample")



lmlist <- list()
arsqlist <- list()
rselist <- list()
plist <- list()
rmselist <- list()


for (i in 1:length(markerlist)){
  ilm <- lm(as.formula(paste0("predage~", markerlist[i])), md)
  lmlist[i] <- ilm
  arsqlist[i]<- summary(ilm)$adj.r.squared
  rmselist[i] <- sqrt(sum(residuals(ilm)^2) / df.residual(ilm))
  
  
}


lmdf <- data.frame(name =markerlist, rsq=unlist(arsqlist), rmse=unlist(rmselist)) %>%
  left_join(geneinf, by="name")


lmdf$sname <- substr(lmdf$name, 1,2)
lmdf$sname <- factor(lmdf$sname, levels=lmdf$sname[order(lmdf$rsq)])


md <- t(subset(t(ctall.norm), sampleinfo$sex == "F" &
               sampleinfo$qual == "ok"& 
 #                sampleinfo$exp %in% c("OvY", "Size-age", "cohorts") & 
                 sampleinfo$ctg>2 & 
   sampleinfo$ctg <5 &
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")&
 !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))
  
  
  ))

si <- subset(sampleinfo, sampleinfo$sample %in% colnames(md))
old <- t(subset(t(md), si$age == "O"))
young <- t(subset(t(md), si$age == "Y"))
#regen <- t(subset(t(ctg1to9), sampleinfo$type == "OR" & sampleinfo$qual != "bad"))



oy <- cbind(old, young)
t.result <-  apply(oy, 1, function (x) t.test(x[1:ncol(old)],x[ncol(old)+1: ncol(oy)]))


pvals <- data.frame(name=rownames(md), pvalovy = unlist(lapply(t.result, function(x) x$p.value)))



lmdf2 <- left_join(lmdf, pvals, by="name")



mnmd <- gather(nmd, key="sample", value="normexp", ns)


mnmd <- left_join(mnmd, sampleinfo, by="sample")

mnmd<- subset(mnmd, (!is.na(predage))) %>%
  left_join(select(minf, sname, chosenand), by="sname") %>%
  left_join(select(lmdf2, name, rsq, rmse, pvalovy), by="name")

mnmd$grsq <- mnmd$rsq > 0.06

mnmd$ca <- ifelse(mnmd$chosenand == T, "amongst chosen 50", "not amongst chosen 50")

ggplot(mnmd, aes(x=predage, y=normexp))+
#  geom_point(shape=21, aes(fill=type))+
#  geom_text(aes(label=sample, colour=type))+
   geom_smooth(method="lm", se=F, aes(size=sname, colour=grsq),alpha=0.3,linetype=1)+
  scale_size_manual(values=rep(0.5,195), guide=F)+
  theme_bw()+
#scale_y_continuous(trans="log2")+
 scale_fill_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  ylab("Normalised Expression") +
#  xlab('pre')+
  facet_grid(dir~ca)+
  scale_colour_manual(values=c("grey", "forest green"))+
  xlab("Sample age")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
```






#Old Random Forest model

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
mymodel <- trterfagemodel
md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
#               sampleinfo$qual == "ok"&
                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") &
 #                sampleinfo$ctg>2 &
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")
))

md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])
#md <- subset(md, rownames(md) %in% rownames(mymodel$trainingData))

md <- md[,colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)



md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample")

md <- subset(md, !(is.na(md$predage)))


RMSEtr <- sqrt(mean( (md$rfagepred[md$intrain == T] - md$predage[md$intrain == T])   ^2))





RMSEte <- sqrt(mean( (md$rfagepred[md$intrain == F] - md$predage[md$intrain == F])   ^2))

(p=ggplot(md, aes(x=predage, y=rfagepred,colour=intrain))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0(length(mymodel$coefnames), 
            " markers in model \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3))))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )#+  facet_wrap(~intrain)




````

#Quality assurance 1: Read number

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


md <- as.data.frame(t(subset(t(ctall.unnorm), 
                 sampleinfo$prep == "Nikos" & 
                 sampleinfo$ctg %in% 3:4&
  !(sampleinfo$sample %in% c("O1a", "O1b", "YR1a", "YR1b" , "YR2a", "YR2b")))))


si <- subset(sampleinfo, sampleinfo$sample %in% colnames(md))


sl <- as.character(sampleinfo$sample[sampleinfo$sample %in% colnames(md)])

md$marker <- rownames(md)

mmd <- gather(as.data.frame(md), key="sample", value="expression", sl) 
mmd$gp <- substr(mmd$sample, 1, nchar(mmd$sample)-1)

mmd$rep <- substr(mmd$sample,  nchar(mmd$sample), nchar(mmd$sample))


md <- select(mmd, -sample) 


mmd2 <- left_join( select(dplyr::rename(filter(md, rep=="a"), expression_a =expression), -rep),
                   select(dplyr::rename(filter(md, rep=="b"), expression_b =expression), -rep),
                   by=c("marker", "gp") )


ggplot(mmd2, aes(x=expression_a, y=expression_b))+
  geom_point(alpha=0.1, colour="red2")+
  geom_smooth(method="lm")+
# geom_hex()+
  theme_bw()+
   scale_x_log10()+
  scale_y_log10()+
  xlab("Expression Replicate B")+
  ylab("Expression Replicate A")+
 ggtitle("How much expression needed for replicability? ", subtitle = "All genes, samples with replicates")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent"))

````


#Quality assurance 2: Housekeeping gene ratio


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

log <-    sampleinfo$flag == FALSE


mhk <- as.data.frame(t(ctall.hk[,log==T])) 

expected <- colSums(mhk)/sum(colSums(mhk))

mhkall <-  as.data.frame(t(ctall.hk)) 

mhkall <- subset(mhkall, rowSums(mhkall) !=0)

eo <- cbind(mhkall, t(data.frame(expected)))

   # expected proportions


chi.result <- apply(eo, 1, function (x) chisq.test(x=x[1:5],p=x[6:10]))
mhkall$p_value <- unlist(lapply(chi.result, function(x) x$p.value))

mhkall$chistat <- unlist(lapply(chi.result, function(x) x$statistic))

mhkall$flag <- rownames(mhkall) %in% sampleinfo$sample[sampleinfo$flag==T]
mhkall$sample <- rownames(mhkall)




mhkall$sample <- factor(mhkall$sample, levels=mhkall$sample[order(mhkall$chistat)])


(p=ggplot(mhkall, aes(y=chistat,x=sample, fill=flag))+
  geom_bar(stat="identity")+
coord_flip()+
  theme_bw()+
  ylab("Chi-square stat (deviation from expected housekeeping ratio)")+
scale_fill_brewer(palette="Set1"))+
  ggtitle("Ranking deviation from observed/expected")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent"))


```

#40-fold cross-validation

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
mymodel <- rfagemodel61
mymodel <- currbest


md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &

                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") &
                 sampleinfo$ctg>2 &
                 sampleinfo$type %in% c("O", "Y", "M") 
)
)
md <- md[, lmdf2$in61 ==T]


md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)



md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample")

md <- subset(md, !(is.na(md$predage)))


RMSEtr <- sqrt(mean( (md$rfagepred[md$intrain == T] - md$predage[md$intrain == T])   ^2))
RMSEte <- sqrt(mean( (md$rfagepred[md$intrain == F] - md$predage[md$intrain == F])   ^2))


(p=ggplot(md, aes(x=predage, y=rfagepred,colour=intrain))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0(length(mymodel$coefnames), 
            " markers in model \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3))))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent"))#+  facet_wrap(~intrain)





#Tuning parameters for a random forest
````


#Contributions to model


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
mymodel <- currbest

vrimprf <-varImp(mymodel)$importance %>%
  rownames_to_column(var="name")
vrimprf$sname <- substr(vrimprf$name, 1,2)

vrimprf$sname <- factor(vrimprf$sname, levels=vrimprf$sname[order(vrimprf$Overall, decreasing=T)])

(p=ggplot(vrimprf, aes(x=sname, y=Overall))+
  theme_bw()+
  geom_bar(stat="identity", fill="red3")+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("Marker")+
  ylab("Contribution to model")+
  ggtitle("Marker contributions", subtitle= paste0("random forest model built with ", length(mymodel$coefnames), " markers"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent"))
)

#ggsave(plot=p,height=5,width=5,dpi=200, filename=paste("contribs.pdf"), useDingbats=FALSE, limitsize = FALSE)
````


#Marker profiles



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.height = 12}


md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
#               sampleinfo$qual == "ok"& 
                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") & 
 #                sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))



md <- md[,colnames(md) %in% mymodel$coefnames]
md <- subset(md, rownames(md) %in% sampleinfo$sample[!is.na(sampleinfo$predage)])

md <-  as.data.frame(md/rowMeans(md))

md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample") %>%
  gather(key=marker, value= exp, mymodel$coefnames)%>%
  left_join(minf, by=c("marker"="name"))%>%
  left_join(select(vrimprf, -sname), by= c("marker"="name"))

md$lab <- paste0(md$sname, ": ", round(md$Overall,3))


md$lab <- factor(md$lab, levels=unique(md$lab[order(md$Overall, decreasing=T)]))

(p=ggplot(md, aes(x=predage, y= exp, colour=dir))+
  geom_smooth()+
  geom_point(size=0.5)+
  theme_bw()+
  facet_wrap(~lab, scales="free", ncol=7)+
  scale_colour_brewer(palette="Set1")+
  xlab("Age (based on body length)")+
  ylab("Normalised expression")+
  theme(legend.direction = 'horizontal', legend.position = 'bottom')+
 
  ggtitle("Marker profiles", subtitle= paste0("markers from random forest model built with ", length(mymodel$coefnames), " markers")) +
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent")))



#ggsave(plot=p,height=18,width=6,dpi=200, filename=paste("contribprofiles.pdf"), useDingbats=FALSE, limitsize = FALSE)



````

#Gradient Boosting Machine

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
mymodel <- rfagemodel61
mymodel <- gbmagemodel


md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &

                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") &
                 sampleinfo$ctg>2 &
                 sampleinfo$type %in% c("O", "Y", "M") 
)
)
md <- md[, lmdf2$in61 ==T]


md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata=md)


md$intrain <- rownames(md) %in% rownames(mymodel$trainingData)



md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample")

md <- subset(md, !(is.na(md$predage)))


RMSEtr <- sqrt(mean( (md$rfagepred[md$intrain == T] - md$predage[md$intrain == T])   ^2))
RMSEte <- sqrt(mean( (md$rfagepred[md$intrain == F] - md$predage[md$intrain == F])   ^2))


ggplot(md, aes(x=predage, y=rfagepred,colour=intrain))+
  geom_point()+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black")+
  theme_bw()+
 scale_colour_manual(values=c("cornflower blue", "red3"), name="Data partition", labels=c("Test", "Training"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age", 
          subtitle=paste0(length(mymodel$coefnames), 
            " markers in model \nRMSE in training data: ", round(RMSEtr,3), 
                          "\nRMSE in test data: ", round(RMSEte,3)))+
  theme(plot.background = element_rect(fill = "transparent",colour = "transparent"))#+  facet_wrap(~intrain)





#Tuning parameters for a random forest
````

#Marker contributions
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.show = 'hide'}


mymodel <- currbestgbm

mymodel <- gbmagemodel
vrimpgbm <- summary(mymodel)

````

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


vrimpgbm$sname <- substr(vrimpgbm$var,1,2)
vrimpgbm$sname <- factor(vrimpgbm$sname, levels=vrimpgbm$sname[order(vrimpgbm$rel.inf, decreasing=T)])

ggplot(vrimpgbm, aes(x=sname, y=rel.inf))+
  theme_bw()+
  geom_bar(stat="identity", fill="red3")+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  xlab("Marker")+
  ylab("Contribution to model")+
  ggtitle("Marker contributions", subtitle= paste0("GBM model built with ", length(mymodel$coefnames), " markers"))+
  theme(plot.background = element_rect(fill = "transparent",colour = "transparent"))


````

#Marker contributions and profiles

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE, fig.height = 12}


md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
#               sampleinfo$qual == "ok"& 
                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") & 
 #                sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))



md <- md[,colnames(md) %in% mymodel$coefnames]
md <- subset(md, rownames(md) %in% sampleinfo$sample[!is.na(sampleinfo$predage)])

md <-  as.data.frame(md/rowMeans(md))

md <- md %>%
  rownames_to_column(var="sample") %>%
  left_join(sampleinfo,  by="sample") %>%
  gather(key=marker, value= exp, mymodel$coefnames)%>%
  left_join(minf, by=c("marker"="name"))%>%
  left_join(select(vrimpgbm, -sname), by= c("marker"="var"))

md$lab <- paste0(md$sname, ": ", round(md$rel.inf,3))


md$lab <- factor(md$lab, levels=unique(md$lab[order(md$rel.inf, decreasing=T)]))

(p=ggplot(md, aes(x=predage, y= exp, colour=dir))+
  geom_smooth()+
  geom_point(size=0.5)+
  theme_bw()+
  facet_wrap(~lab, scales="free", ncol=7)+
  scale_colour_brewer(palette="Set1")+
  xlab("Age (based on body length)")+
  ylab("Normalised expression")+
  theme(legend.direction = 'horizontal', legend.position = 'bottom')+
 
  ggtitle("Marker profiles", subtitle= paste0("markers from random forest model built with ", length(mymodel$coefnames), " markers")) +
  theme(plot.background = element_rect(fill = "transparent",colour = "transparent")))





#ggsave(plot=p,height=18,width=6,dpi=200, filename=paste("gbmcontribprofiles.pdf"), useDingbats=FALSE, limitsize = FALSE)


```


#GBM vs Random Forest

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


vrimp <- left_join(vrimpgbm, vrimprf, by="sname")

ggplot(vrimp, aes(x=rel.inf,y=Overall, label=sname))+
  theme_bw()+
  ylab("Random Forest importance")+
  xlab("GBM relative influence")+
  scale_y_continuous(trans="log2")+
  scale_x_continuous(trans="log2")+
#  scale_y_log10()+
#  scale_x_log10()+
  geom_point(size=3, colour="red2")+
  geom_text_repel()+
  ggtitle("Marker importance: GBM vs Random Forest", subtitle= paste0("Same ", length(mymodel$coefnames), " markers"))+
  theme(plot.background = element_rect(fill = "transparent",colour = "transparent"))


````



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctall.norm), 
                           (sampleinfo$sex == "F" &
                 sampleinfo$exp %in% c("regen") & 
                  sampleinfo$ctg > 5 &
                 sampleinfo$type %in% c("O", "Y", "M", "OR", "MR", "YR")
 )))


mymodel <- currbestgbm

mymodel <- rfagemodel61
md <- md[, colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata= md)



md$sample <- rownames(md)

md <- left_join(md, sampleinfo, by="sample") %>%
  left_join(hkdat, by="sample") 





md$group <- substr(md$sample, 1,1)
md$ty <- substr(md$sample, 2,3)

mdhk <-    select(md, over10, group, ty ) %>%
  spread(value= over10, key=ty)

colnames(mdhk)[2:4] <- paste0(colnames(mdhk)[2:4], "over")


md2 <- select(md, rfagepred, group, ty) %>%
  spread(value=rfagepred,key=ty) %>%
  left_join(mdhk, by= "group")


md2$control <- md2$un
md2$control[md2$beover > md2$unover & !is.na(md2$beover)] <- md2$be[md2$beover > md2$unover & !is.na(md2$beover)]


md2$dir <- md2$un > md2$re
md2$diff <-  md2$re- md2$un
md2$regn <- paste0(md2$group, "reg")

md2 <- left_join(md2, select(sampleinfo, sample, predage), by=c("regn"= "sample")) 


 #left_join(hkdat, sampleinfo, by="sample") %>%
atleast <-10
 overthresh <- 1
 chithresh <- 500

 
 
md2$good <- paste0(md2$group, "un") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "un") %in% qualsum$sample[qualsum$over10 >= overthresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$over10 >= overthresh] 

#md2$good= <-  md2$unover >=thresh & md2$reover >=thresh

limrange <- max(abs(md2$diff))


````

#Does regeneration rejuvenate?


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


sik <- left_join(sampleinfo, hkdat, by="sample")
md <- as.data.frame(subset(t(ctall.norm), 
                           (sik$sex == "F" &
                              sik$over10 >= overthresh&
                 sik$exp %in% c("Size-age", "cohorts", "OvY", "regen") & 
                 sik$type %in% c("O", "Y", "M", "OR", "MR", "YR")
                 &                 !  sik$sample %in% c("AF", "AG")
)) 
 )


md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

mymodel <- currbest


md <- md[, colnames(md) %in% mymodel$coefnames]
md$rfagepred <- predict(mymodel, newdata= md)

md$sample <- rownames(md)

md <- left_join(md, sik, by="sample") %>%
  filter(!(is.na(predage)))

lm <- lm(rfagepred~predage, data=md)
 
reglist <- c(paste0(md2$group[md2$good], c("un")), paste0(md2$group[md2$good], c("reg")))


mlena <- gather(lena, key="model", value="pred", c("bcpred", "linpred", "polypred2", "polypred3"))

# 
mmd2 <- gather(subset(md2, md2$good), value="modelage",key="state", c("re", "un"))



(p=ggplot(md, aes(x=predage, y=rfagepred))+

#  geom_text_repel(aes(label=sample), size=2.5)   + 
  geom_point(colour="grey")+
  xlab("Age (weeks) inferred from body length") +
  ylab("Age (weeks) inferred from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black", data=subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") ))+
  theme_bw()+
  geom_segment(data=subset(md2, md2$good), aes(x=predage, xend=predage, y=un, yend=re, colour=dir),alpha=1, size=1.5)+
    geom_point(data=mmd2, aes(x=predage, y=modelage, fill=state), shape=21)+
  scale_fill_manual(name="", 
                    values=c("lime green", "cornflower blue", "forest green"), 
                    labels=c("Regenerated limb", "Contralateral limb"))+
  scale_colour_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
    ggtitle("Change in molecular age due to regeneration", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
) 
  

```

#Does regeneration rejuvenate?


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
 

md <- as.data.frame(subset(t(ctall.norm), 
                           (sampleinfo$sex == "F" &
                 sampleinfo$exp %in% c("regen") & 
                  sampleinfo$ctg > 5 &
                 sampleinfo$type %in% c("O", "Y", "M", "OR", "MR", "YR")
 )))



mymodel <- currbest

#mymodel <- rfagemodel61
md <- md[, colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata= md)



md$sample <- rownames(md)

md <- left_join(md, sampleinfo, by="sample") %>%
  left_join(hkdat, by="sample") 





md$group <- substr(md$sample, 1,1)
md$ty <- substr(md$sample, 2,3)

mdhk <-    select(md, over10, group, ty ) %>%
  spread(value= over10, key=ty)

colnames(mdhk)[2:4] <- paste0(colnames(mdhk)[2:4], "over")


md2 <- select(md, rfagepred, group, ty) %>%
  spread(value=rfagepred,key=ty) %>%
  left_join(mdhk, by= "group")


md2$control <- md2$un
md2$control[md2$beover > md2$unover & !is.na(md2$beover)] <- md2$be[md2$beover > md2$unover & !is.na(md2$beover)]


md2$dir <- md2$un > md2$re
md2$diff <-  md2$re- md2$un
md2$regn <- paste0(md2$group, "reg")

md2 <- left_join(md2, select(sampleinfo, sample, predage), by=c("regn"= "sample")) 


 #left_join(hkdat, sampleinfo, by="sample") %>%
atleast <-10
 overthresh <- 1
 chithresh <- 500

 
 
md2$good <- paste0(md2$group, "un") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "un") %in% qualsum$sample[qualsum$over10 >= overthresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$over10 >= overthresh] 

#md2$good= <-  md2$unover >=thresh & md2$reover >=thresh

limrange <- max(abs(md2$diff))


#round(wilcox.test(md2$un[md2$good], md2$re[md2$good])$p.value,3)

md2$group <- factor(md2$group, levels=md2$group[order(md2$diff)])

(p=
    ggplot(subset(md2, md2$good), aes(x=group, y=diff, fill=dir)) +
  ylim(-limrange, limrange)+
  theme_bw()+
  ylab("Rejuvenation Effect (weeks age difference)")+
  xlab("Sample Group")+
  geom_bar(stat="identity") +
  scale_fill_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
  ggtitle("Rejuvenation Effect", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used", "\np-value of t-test between regenerated and unamputated: ", 
round(t.test(md2$un[md2$good], md2$re[md2$good])$p.value,3))) 


   +
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
) 
  
  
```


#Does regeneration rejuvenate?

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctall.norm), 
                           (sik$sex == "F" &
                              sik$over10 >= overthresh&
                 sik$exp %in% c("Size-age", "cohorts", "OvY", "regen") & 
                 sik$type %in% c("O", "Y", "M", "OR", "MR", "YR")
                 &                 !  sik$sample %in% c("AF", "AG")
)) 
 )


md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

mymodel <-gbmagemodel


md <- md[, colnames(md) %in% mymodel$coefnames]
md$rfagepred <- predict(mymodel, newdata= md)

md$sample <- rownames(md)

md <- left_join(md, sik, by="sample") %>%
  filter(!(is.na(predage)))

lm <- lm(rfagepred~predage, data=md)
 
reglist <- c(paste0(md2$group[md2$good], c("un")), paste0(md2$group[md2$good], c("reg")))


mlena <- gather(lena, key="model", value="pred", c("bcpred", "linpred", "polypred2", "polypred3"))

# 
mmd2 <- gather(subset(md2, md2$good), value="modelage",key="state", c("re", "un"))



(p=ggplot(md, aes(x=predage, y=rfagepred))+

#  geom_text_repel(aes(label=sample), size=2.5)   + 
  geom_point(colour="grey")+
  xlab("Age (weeks) inferred from body length") +
  ylab("Age (weeks) inferred from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black", data=subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") ))+
  theme_bw()+
  geom_segment(data=subset(md2, md2$good), aes(x=predage, xend=predage, y=un, yend=re, colour=dir),alpha=1, size=1.5)+
    geom_point(data=mmd2, aes(x=predage, y=modelage, fill=state), shape=21)+
  scale_fill_manual(name="", 
                    values=c("lime green", "cornflower blue", "forest green"), 
                    labels=c("Regenerated limb", "Contralateral limb"))+
  scale_colour_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
    ggtitle("Change in molecular age due to regeneration", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
) 
  

  
```

#Does regeneration rejuvenate?



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
 

md <- as.data.frame(subset(t(ctall.norm), 
                           (sampleinfo$sex == "F" &
                 sampleinfo$exp %in% c("regen") & 
                  sampleinfo$ctg > 5 &
                 sampleinfo$type %in% c("O", "Y", "M", "OR", "MR", "YR")
 )))



mymodel <- gbmagemodel

#mymodel <- rfagemodel61
md <- md[, colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata= md)



md$sample <- rownames(md)

md <- left_join(md, sampleinfo, by="sample") %>%
  left_join(hkdat, by="sample") 





md$group <- substr(md$sample, 1,1)
md$ty <- substr(md$sample, 2,3)

mdhk <-    select(md, over10, group, ty ) %>%
  spread(value= over10, key=ty)

colnames(mdhk)[2:4] <- paste0(colnames(mdhk)[2:4], "over")


md2 <- select(md, rfagepred, group, ty) %>%
  spread(value=rfagepred,key=ty) %>%
  left_join(mdhk, by= "group")


md2$control <- md2$un
md2$control[md2$beover > md2$unover & !is.na(md2$beover)] <- md2$be[md2$beover > md2$unover & !is.na(md2$beover)]


md2$dir <- md2$un > md2$re
md2$diff <-  md2$re- md2$un
md2$regn <- paste0(md2$group, "reg")

md2 <- left_join(md2, select(sampleinfo, sample, predage), by=c("regn"= "sample")) 


 #left_join(hkdat, sampleinfo, by="sample") %>%
atleast <-10
 overthresh <- 1
 chithresh <- 500

 
 
md2$good <- paste0(md2$group, "un") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "un") %in% qualsum$sample[qualsum$over10 >= overthresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$over10 >= overthresh] 

#md2$good= <-  md2$unover >=thresh & md2$reover >=thresh

limrange <- max(abs(md2$diff))


#round(wilcox.test(md2$un[md2$good], md2$re[md2$good])$p.value,3)

md2$group <- factor(md2$group, levels=md2$group[order(md2$diff)])

(p=
    ggplot(subset(md2, md2$good), aes(x=group, y=diff, fill=dir)) +
  ylim(-limrange, limrange)+
  theme_bw()+
  ylab("Rejuvenation Effect (weeks age difference)")+
  xlab("Sample Group")+
  geom_bar(stat="identity") +
  scale_fill_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
  ggtitle("Rejuvenation Effect", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used", "\np-value of t-test between regenerated and unamputated: ", 
round(t.test(md2$un[md2$good], md2$re[md2$good])$p.value,3))) 


   +
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
) 
  
  
```


