---
title: "Lab Meeting"
author: "Luke Hayden"
date: 26th September 2018
output:
  beamer_presentation:
    theme: "Darmstadt"
    colortheme: "fly"
    fonttheme: "default"
slide_level: 3
---



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(ggplot2)
library(rmarkdown)
library(RColorBrewer)
library(scales)
library(ggbiplot)
library(devtools)
library(Rmisc)
library(dplyr)
library(caret)
library(RColorBrewer)
library(ggrepel)
library(gtools)
library(FinCal)
library(tidyr)
library(tibble)


load(file="nov17main_data.rdata")

load(file="allns_data.rdata")
load(file="markerchoiceinfo.rdata")

load(file="rfagemodel.rdata")

load(file="trterfagemodel.rdata")
```



#Background
##Project

Examine the effect of regeneration on the molecular age profile of *Parhyale* limbs

##Designing codeset

*Nanostring as method to quantify gene expression

*200 genes in codeset

-195 genes chosen on the basis of differential expression analysis 

-5 control genes: do not vary in expression between conditions


#Age-length relationship

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


zklen <- read.csv("zk-age-len.csv")[2:3] %>%
  mutate(prep="ZK")

lena <- sampleinfo %>%
  filter(!is.na(wkage)) %>%
  select(length, wkage, prep) %>%
  rbind(zklen)



ggplot(lena, aes(y=length, x=wkage))+
  geom_point()+
  geom_smooth()+
  theme_bw()+
  scale_colour_brewer(palette="Set1")+
  ggtitle("Age-length relationship in female Parhyale")+
  ylab("Body length (mm)")+
  xlab("Age (weeks)")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )



````

#Building an age-length model



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

lena$lensq <- (lena$length)^2

lena$wkage <- as.numeric(lena$wkage)

sqmodel <- train(
  wkage~lensq, lena,
   method = "lm",
  trControl = trainControl(
    method = "cv", number = 5,
    repeats = 5, verboseIter = F
  )
)

#sqmodel

lena$sqpred <- predict(sqmodel)
lena$sqresid <- lena$wkage -lena$sqpred

#mlena <- gather(lena,key="model", value="pred", c("sqpred", "linpred", "logpred", "rfpred"))



ggplot(lena, aes(x=wkage, y=sqpred))+
  geom_point()+
 # facet_wrap(~model, ncol=2)+
  theme_bw()+
  scale_colour_brewer(palette="Set1")+
  ggtitle("Model predicted vs real values", subtitle="")+
  xlab("Age (weeks): predicted")+
  ylab("Age (weeks): real data")+geom_smooth(method="lm")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )

````


#Young vs old separation: PCA old vs young

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



log <- sampleinfo$reg == "" & 
  !(sampleinfo$age %in% c("N")) & 
  sampleinfo$sex == "F"&
  sampleinfo$exp %in% c("OvY", "Size-age") & 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8")) &
 sampleinfo$over10reads > 100 &
  sampleinfo$ctg > 2 &
  sampleinfo$prep == "Luke"

bcmin <- as.matrix(ctallgood.norm[,log== TRUE])
bcmin.groups <- subset(sampleinfo$type, log==TRUE)

tbc <- as.matrix(t(bcmin))

bc.pca <- prcomp(tbc,center = TRUE,scale. = TRUE) 
bc.groups <- bcmin.groups


ggbiplot(bc.pca, obs.scale = 1, var.scale = 1, ellipse = FALSE, circle = TRUE, var.axes=F, labels=rownames(tbc), groups=bc.groups)+ggtitle("PCA of Old vs Young (females): 50 chosen markers")+
    theme_bw() +
#  scale_colour_manual(values=c("cornflower blue", "red3", "black"))+
  theme(plot.title=element_text(size=8,face="bold"))+
#  geom_label(label=rownames(tbc), aes(colour=bc.groups))+
  scale_colour_manual(values= c("darkviolet","cornflower blue",  "red2"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )


````

#Young vs old separation: Old vs young by marker

```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- t(subset(t(ctall.norm), sampleinfo$sex == "F" &
  #              sampleinfo$qual == "ok"& 
                 sampleinfo$exp %in% c("OvY", "Size-age", "cohorts") & 
                 sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))

si <- subset(sampleinfo, sampleinfo$sample %in% colnames(md))
old <- t(subset(t(md), si$age == "O"))
young <- t(subset(t(md), si$age == "Y"))
#regen <- t(subset(t(ctg1to9), sampleinfo$type == "OR" & sampleinfo$qual != "bad"))





nmd <- as.data.frame(md/rowMeans(md))


ns <- colnames(md)

nmd$name <- rownames(md)
nmd$sname <- substr(nmd$name, 1,2)

oy <- cbind(old, young)
t.result <-  apply(oy, 1, function (x) t.test(x[1:ncol(old)],x[ncol(old)+1: ncol(oy)]))



nmd$p_value <- unlist(lapply(t.result, function(x) x$p.value))

nmd <- left_join(nmd, geneinf, by="name")






nmd$YvOl2fc <- foldchange2logratio(foldchange(rowMeans(young), rowMeans(old)))


nmd$sname <- factor(nmd$sname, levels=nmd$sname[order(nmd$YvOl2fc)])
nmd$good <- nmd$p_value <0.05
nmd$di <- paste0(ifelse(nmd$dir == "upregulated with age", "up", "down"), ": ", nmd$sname)


sl <- as.character(sampleinfo$sample[sampleinfo$sample %in% colnames(nmd)])

mmd <- gather(as.data.frame(md), key="sample", value="expression", sl)

mnmd <- gather(nmd, key="sample", value="normexp", sl) %>%
  left_join(sampleinfo, by= "sample") %>%
  left_join(mmd, by="sample")


ggplot(mnmd, aes(x= sname, y=normexp, colour=type))+
#  geom_point()+
  theme_bw()+
  geom_boxplot(outlier.shape=NA)+
  scale_y_continuous(trans="log2")+
  facet_grid(~dir, space="free",scales="free" )+
#  geom_boxplot(outlier.shape = NA)+
  scale_colour_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  scale_fill_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  ylab("Normalised Expression") +
  xlab('Marker')+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
  


````

#Expression/length relationship


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctall.norm), sampleinfo$sex == "F" &
  #              sampleinfo$qual == "ok"& 
#                 sampleinfo$exp %in% c("OvY", "Size-age", "cohorts") & 
                 sampleinfo$ctg>2 & 
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")))
 # !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))))



markerlist <- colnames(md)

md$sample <- rownames(md)


md <- left_join(md, select(sampleinfo, sample, predage), by="sample") %>%
  filter(!(is.na(predage))) %>%
  column_to_rownames(var="sample")



lmlist <- list()
arsqlist <- list()
rselist <- list()
plist <- list()
rmselist <- list()


for (i in 1:length(markerlist)){
  ilm <- lm(as.formula(paste0("predage~", markerlist[i])), md)
  lmlist[i] <- ilm
  arsqlist[i]<- summary(ilm)$adj.r.squared
  rmselist[i] <- sqrt(sum(residuals(ilm)^2) / df.residual(ilm))
  
  
}


lmdf <- data.frame(name =markerlist, rsq=unlist(arsqlist), rmse=unlist(rmselist)) %>%
  left_join(geneinf, by="name")


lmdf$sname <- substr(lmdf$name, 1,2)
lmdf$sname <- factor(lmdf$sname, levels=lmdf$sname[order(lmdf$rsq)])


md <- t(subset(t(ctall.norm), sampleinfo$sex == "F" &
               sampleinfo$qual == "ok"& 
 #                sampleinfo$exp %in% c("OvY", "Size-age", "cohorts") & 
                 sampleinfo$ctg>2 & 
   sampleinfo$ctg <5 &
                sampleinfo$codeset == "phaw_1" &
                 sampleinfo$type %in% c("O", "Y", "M")&
 !(sampleinfo$sample %in% c("T","C", "U", "LFPD", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))
  
  
  ))

si <- subset(sampleinfo, sampleinfo$sample %in% colnames(md))
old <- t(subset(t(md), si$age == "O"))
young <- t(subset(t(md), si$age == "Y"))
#regen <- t(subset(t(ctg1to9), sampleinfo$type == "OR" & sampleinfo$qual != "bad"))



oy <- cbind(old, young)
t.result <-  apply(oy, 1, function (x) t.test(x[1:ncol(old)],x[ncol(old)+1: ncol(oy)]))


pvals <- data.frame(name=rownames(md), pvalovy = unlist(lapply(t.result, function(x) x$p.value)))



lmdf2 <- left_join(lmdf, pvals, by="name")



mnmd <- gather(nmd, key="sample", value="normexp", ns)


mnmd <- left_join(mnmd, sampleinfo, by="sample")

mnmd<- subset(mnmd, (!is.na(predage))) %>%
  left_join(select(minf, sname, chosenand), by="sname") %>%
  left_join(select(lmdf2, name, rsq, rmse, pvalovy), by="name")

mnmd$grsq <- mnmd$rsq > 0.06

mnmd$ca <- ifelse(mnmd$chosenand == T, "amongst chosen 50", "not amongst chosen 50")

ggplot(mnmd, aes(x=predage, y=normexp))+
#  geom_point(shape=21, aes(fill=type))+
#  geom_text(aes(label=sample, colour=type))+
   geom_smooth(method="lm", se=F, aes(size=sname, colour=grsq),alpha=0.3,linetype=1)+
  scale_size_manual(values=rep(0.5,195), guide=F)+
  theme_bw()+
#scale_y_continuous(trans="log2")+
 scale_fill_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  theme(axis.text.x = element_text(angle = 60, hjust = 1))+
  ylab("Normalised Expression") +
#  xlab('pre')+
  facet_grid(dir~ca)+
  scale_colour_manual(values=c("grey", "forest green"))+
  xlab("Sample age")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
```


#Multiple Regression approach
####Simple linear regression: 

**Age = X(marker1) + c**

We try to find values for x & c that come as close as possible to solving the equation for each set of values for *Age* and *marker1* we have. 

####Two predictors:

**Age = X(marker1) + Y(marker2) + c**

####Many predictors

**Age = X(marker1) + Y(marker2) + Z(marker3) + W(marker4) + .... + c**


Where we have many different markers, we can find values of x,y,z,w, etc that solve this equation very well but don't provide predictive power

#Random Forest approach

###Decision tree

Classify or perform regression by asking binary questions of data: whether value of marker X is above or below key value Y, whther marker Z is above or below.....

###Random Forest

Ensemble of decision trees, each using a random subset of the predictors

Resists overfitting

#Examining this model



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- modelsamp

md$rfagepred <- predict(rfagemodel)

#md <- md %>%
#  rownames_to_column(var="sample") %>%
 # left_join(select(sampleinfo, -predage), by="sample")


accrf <- lm(predage~rfagepred,md)


ggplot(md, aes(x=predage, y=rfagepred))+
  geom_point(aes(colour=type))+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black")+
  theme_bw()+
 scale_colour_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )


```


#Out-of-sample errors 


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
olddat <- trterfagemodel$trainingData


md <- as.data.frame(subset(t(ctall.norm), 
                           (sampleinfo$sex == "F" &
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") & 
               sampleinfo$codeset == "phaw_1" &
                  sampleinfo$qual != "bad"&
 #                 sampleinfo$over10reads >40 &
                 sampleinfo$type %in% c("O", "Y", "M", "OR", "MR", "YR")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD","Areg", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))) 
 ))

md <- md[, colnames(md) %in% colnames(olddat)]

#md <-modelsamptrte

md$intrain <- rownames(md) %in% rownames(olddat)


md$trterfagepred <- predict(trterfagemodel, newdata=md)

md$sample <- rownames(md)


md <- left_join(md, select(sampleinfo, sample, predage), by="sample") %>%
  filter(!(is.na(predage))) 

accrf <- lm(predage~trterfagepred, subset(md, md$intrain==F))


ggplot(md, aes(x=predage, y=trterfagepred))+
  geom_point(aes(colour=intrain))+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black")+
  theme_bw()+
# scale_colour_manual(values=c("darkviolet", "cornflower blue", "red3"))+
  geom_text_repel(aes(label=sample), size=2.5)+
  ggtitle("Using Marker gene expression to predict age")+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )


```

#Effect of regeneration



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctall.norm), 
                           (sampleinfo$sex == "F" &
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY", "regen", "regen-long") & 
               sampleinfo$codeset == "phaw_1" &
                  sampleinfo$qual != "bad"&
 #                 sampleinfo$over10reads >40 &
                 sampleinfo$type %in% c("O", "Y", "M", "OR", "MR", "YR")& 
  !(sampleinfo$sample %in% c("T","C", "U", "LFPD","Areg", "O1a", "SFPA", "LFIB","LFIA","AF", "Coh8"))) 
 ))


md <- md[,colnames(md) %in% colnames(modelsamp)]

md$rfagepred <- predict(rfagemodel, newdata= md)




md$sample <- rownames(md)

md <- left_join(md, sampleinfo, by="sample") %>%
  filter(!(is.na(predage)))

md$type <- as.character(md$type)
md$type <- as.factor(case_when(
  md$type %in% c("O", "M", "Y") ~ as.character(md$type), 
  md$type %in% c("MR", "YR", "OR", "OR-A") ~"R"
))


md$regentime[is.na(md$regentime)] <- ""


md$typer <- paste0(md$type, md$regentime)


lm <- lm(rfagepred~predage, data=md)
 
#lm <- lm(rfagepred~predage, data=subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") ))
                 
md$lmresid <- lm$residuals
md$lmressign <- md$lmresid >0

(p=ggplot(md, aes(x=predage, y=rfagepred))+
  geom_point(aes(colour=type))+
  xlab("Age (weeks) from body length") +
  ylab("Age (weeks) from marker gene expression (random forest model)")+
#  geom_line(aes(predage, lm
#  geom_smooth(method="lm")+
  geom_smooth(method="lm", colour="black", data=subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") ))+
  theme_bw()+
    scale_colour_manual(values=c("darkviolet", "cornflower blue", "forest green",  "red2"), name="")+
  geom_text_repel(aes(label=sample,colour=type), size=2.5)+
  ggtitle("How does regeneration affect molecular age?"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
````


#Testing this effect

Variably significant or not depending on details of model (choice of markers used, etc)
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}



t.test(md$lmresid[md$reg =="R"] , alternative="less" )
  

````




#New questions

###How to optimise model?
Need to reduce out of sample error

###Better way of ensuring sample quality?




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


  

````


