---
title: "Regeneration effect"
author: "Luke Hayden"
output: html_document
---


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(dplyr)
#library(ggbiplot)
library(tibble)
library(tidyr)
library(caret)
library(RColorBrewer)
library(ggrepel)
library(gtools)
#library(FinCal)
library(ggrepel)
#library(e10)
````



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
#setwd("~/Documents/nstringjul18/nstringfull")
load(file="allns_data.rdata")
load(file="modelrf.rdata")
#load(file="modelgbm.rdata")
load(file="hkdat.rdata")
load(file="sampleinfo.rdata")
load(file="qualitydat.rdata")

load("fullns_data.rdata")
load(file="qualitydat.rdata")
load(file="qualitydatwnew.rdata")


sik <- left_join(sampleinfob, hkdat, by="sample")

````



#Regenerated vs controls


```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctfull.norm), 
                           (sampleinfo$sex == "F" &
                 sampleinfo$exp %in% c("regen") & 
                  sampleinfo$ctg > 5 &
                 sampleinfo$type %in% c("O", "Y", "M", "OR", "MR", "YR")
 )))



mymodel <- gbmrsqopt
md <- md[, colnames(md) %in% mymodel$coefnames]



md$rfagepred <- predict(mymodel, newdata= md)



md$sample <- rownames(md)

md <- left_join(md, sampleinfob, by="sample") %>%
  left_join(hkdat, by="sample") 





md$group <- substr(md$sample, 1,1)
md$ty <- substr(md$sample, 2,3)

mdhk <-    select(md, over10, group, ty ) %>%
  spread(value= over10, key=ty)

colnames(mdhk)[2:4] <- paste0(colnames(mdhk)[2:4], "over")


md2 <- select(md, rfagepred, group, ty) %>%
  spread(value=rfagepred,key=ty) %>%
  left_join(mdhk, by= "group")


md2$control <- md2$un
md2$control[md2$beover > md2$unover & !is.na(md2$beover)] <- md2$be[md2$beover > md2$unover & !is.na(md2$beover)]


md2$dir <- md2$un > md2$re
md2$diff <-  md2$re- md2$un
md2$regn <- paste0(md2$group, "reg")

md2 <- left_join(md2, select(sampleinfo, sample, predage), by=c("regn"= "sample")) 


 #left_join(hkdat, sampleinfo, by="sample") %>%
atleast <-10
 overthresh <- 1
 chithresh <- 500

 
 
md2$good <- paste0(md2$group, "un") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "un") %in% qualsum$sample[qualsum$over10 >= overthresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$chistat < chithresh] &
            paste0(md2$group, "reg") %in% qualsum$sample[qualsum$over10 >= overthresh] 

#md2$good= <-  md2$unover >=thresh & md2$reover >=thresh

limrange <- max(abs(md2$diff))


round(wilcox.test(md2$un[md2$good], md2$re[md2$good])$p.value,3)


round(t.test(md2$un[md2$good], md2$re[md2$good], alternative = "less")$p.value,3)


md2$group <- factor(md2$group, levels=md2$group[order(md2$diff)])

ggplot(subset(md2, md2$good), aes(x=group, y=diff, fill=dir)) +
  ylim(-limrange, limrange)+
  theme_bw()+
  ylab("Rejuvenation Effect (weeks age difference)")+
  xlab("Sample Group")+
  geom_bar(stat="identity") +
  scale_fill_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
  ggtitle("Rejuvenation Effect", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used", "\nSamples with at least ", overthresh, " housekeeping genes with at least ",atleast, " reads\nChi-stat < ", chithresh, "\np-value of t-test between regenerated and unamputated: ", 
round(t.test(md2$un[md2$good], md2$re[md2$good])$p.value,3))) 


goodgroups<- subset(md2, md2$good)$group



```

#Regeneration effect visualised on aging time graph



```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}

md <- as.data.frame(subset(t(ctfull.norm), sampleinfo$sex == "F" &
#               sampleinfo$qual == "ok"&
                 sampleinfo$prep== "Luke"&
                 sampleinfo$exp %in% c("Size-age", "cohorts", "OvY") &

                 sampleinfo$type %in% c("O", "Y", "M")
))

md <- subset(md, rownames(md)%in% qualsum$sample[qualsum$good])

md <- md[,colnames(md) %in% mymodel$coefnames]

#mymodel <- currbestgbm


md$rfagepred <- predict(mymodel, newdata= md)

md$sample <- rownames(md)

md <- left_join(md, sik, by="sample") 

#%>%
  filter(!(is.na(predage)))

lm <- lm(rfagepred~predage, data=md)
 
reglist <- c(paste0(md2$group[md2$good], c("un")), paste0(md2$group[md2$good], c("reg")))

# 
mmd2 <- gather(subset(md2, md2$good), value="modelage",key="state", c("re", "un"))



(p=ggplot(md, aes(x=predage, y=rfagepred))+

#  geom_text_repel(aes(label=sample), size=2.5)   + 
  geom_point(colour="grey")+
  xlab("Age (weeks) inferred from body length") +
  ylab("Age (weeks) inferred from marker gene expression (random forest model)")+
  geom_smooth(method="lm", colour="black", data=subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") ))+
  theme_bw()+
  geom_segment(data=subset(md2, md2$good), aes(x=predage, xend=predage, y=un, yend=re, colour=dir),alpha=1, size=1.5)+
    geom_point(data=mmd2, aes(x=predage, y=modelage, fill=state), shape=21)+
  scale_fill_manual(name="", 
                    values=c("lime green", "cornflower blue", "forest green"), 
                    labels=c("Regenerated limb", "Contralateral limb"))+
  scale_colour_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
    ggtitle("Change in molecular age due to regeneration", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used"))+
  theme(
    plot.background = element_rect(fill = "transparent",colour = "transparent") # bg of the panel
  )
) 
  
```




```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}


(p=ggplot(md, aes(x=predage, y=rfagepred))+

#  geom_text_repel(aes(label=sample), size=2.5)   + 
  geom_point(colour="grey")+
  xlab("Age (weeks) from body length") +
  ylab(paste0("Age (weeks) from marker gene expression (", toupper(mymodel$method), ")"))+
  geom_smooth(method="lm", colour="black", data=subset(md,md$exp %in% c("Size-age", "cohorts", "OvY") ))+
  theme_bw()+
  geom_segment(data=subset(md2, md2$good), aes(x=predage, xend=predage, y=un, yend=re, colour=dir),alpha=1, size=1.5)+
  geom_point(data=subset(md2, md2$good), aes(x=predage, y=un),size=3, colour="cornflower blue")+
  geom_point(data=subset(md2, md2$good), aes(x=predage, y=re),size=3, colour="lime green")+
  
  geom_text(data=subset(md2, md2$good), aes(x=predage, y=un), colour="white",size=2, label="C")+
  geom_text(data=subset(md2, md2$good), aes(x=predage, y=re), colour="white",size=2, label="R")+
 # geom_text_repel(data=subset(md, md$sample %in% reglist), aes(label=sample))+
  scale_colour_manual(name="", 
                    values=c("orangered", "forest green"), 
                    labels=c("Aged", "Rejuvenated"))+
  ggtitle("Change in molecular age due to regeneration", 
          subtitle= paste0("Model type: ", toupper(mymodel$method), ", ", length(mymodel$coefnames), " markers used", "\nSamples with at least ", thresh, " housekeeping genes expressed in both  \np-value of t-test between regenerated and unamputated: ", 
round(t.test(md2$un[md2$good], md2$re[md2$good])$p.value,3))) )


#ggsave(plot=p, file="rejuveff.pdf", height=8, width=8)
```


#Regeneration time effect
```{r, echo=FALSE, warning=FALSE, error=FALSE, message=FALSE}
#md$regentime <- as.numeric(md$regentime)
#md$regentime[is.na(md$regentime)] 
md$regentime[md$regentime== ""] <- "unregenerated samples"
md$regentime <- factor(md$regentime, levels= c("unregenerated samples", "4", "7", "13"))

mds <- subset

p=ggplot(md, aes(x=regentime, y=lmresid, fill=regentime))+
  geom_boxplot(outlier.shape=NA)+
 scale_fill_manual(values=c("grey", "aquamarine3","forest green", "limegreen",  "red2"), name="")+
  ylab("Residual variation in molecular age")+
  xlab("Time after regeneration (weeks)")+theme_bw()

ggsave(plot=p, file="regtimeeff.pdf")

```
